src/posix/platform/netif.cpp

```diff
@@ -596,18 +1319,179 @@ void platformNetifInit(otInstance *aInstance, const char *aInterfaceName)
+
+    socklen_t devNameLen;
+    devNameLen = (socklen_t)deviceNameLen;
+    err        = getsockopt(sTunFd, SYSPROTO_CONTROL, UTUN_OPT_IFNAME, deviceName, &devNameLen);
+    VerifyOrDie(err == 0, OT_EXIT_ERROR_ERRNO);
+
+    otLogInfoPlat("Tunnel device name = '%s'", deviceName);
+}
+#endif
+
+#if defined(__NetBSD__) || defined(__FreeBSD__)
+static otError destroyTunnel(void)
+{
+    otError      error;
+    struct ifreq ifr;
+
+    memset(&ifr, 0, sizeof(ifr));
+    strncpy(ifr.ifr_name, sTunName, sizeof(ifr.ifr_name));
+    VerifyOrExit(ioctl(sIpFd, SIOCIFDESTROY, &ifr) == 0, perror("ioctl"); error = OT_ERROR_FAILED);
+    error = OT_ERROR_NONE;
+
+exit:
+
+    return error;
+}
+#endif
+
+#if defined(__NetBSD__) || (defined(__APPLE__) && USE_TUN) || defined(__FreeBSD__)
+static void platformConfigureTunDevice(otInstance *aInstance,
+                                       const char *aInterfaceName,
+                                       char *      deviceName,
+                                       size_t      deviceNameLen)
+{
+    int         flags = IFF_BROADCAST | IFF_MULTICAST;
+    int         err;
+    const char *last_slash;
+    const char *path;
+
+    (void)aInterfaceName;
+    (void)aInstance;
+
+    path = OPENTHREAD_POSIX_TUN_DEVICE;
+
+    sTunFd = open(path, O_RDWR | O_NONBLOCK);
+    VerifyOrDie(sTunFd >= 0, OT_EXIT_ERROR_ERRNO);
+
+#if defined(__NetBSD__) || defined(__FreeBSD__)
+    err = ioctl(sTunFd, TUNSIFMODE, &flags);
+    VerifyOrDie(err == 0, OT_EXIT_ERROR_ERRNO);
+#endif
+
+    flags = 1;
+    err   = ioctl(sTunFd, TUNSIFHEAD, &flags);
+    VerifyOrDie(err == 0, OT_EXIT_ERROR_ERRNO);
+
+    last_slash = strrchr(OPENTHREAD_POSIX_TUN_DEVICE, '/');
+    VerifyOrDie(last_slash != NULL, OT_EXIT_ERROR_ERRNO);
+    last_slash++;
+
+    strncpy(deviceName, last_slash, deviceNameLen);
+}
+#endif
+
+static void platformConfigureNetLink(void)
+{
+#if defined(__linux__)
+    sNetlinkFd = socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+#elif defined(__APPLE__) || defined(__NetBSD__) || defined(__FreeBSD__)
+    sNetlinkFd = socket(PF_ROUTE, SOCK_RAW, 0);
+#else
+#error "!! Unknown platform !!"
+#endif
+    VerifyOrDie(sNetlinkFd >= 0, OT_EXIT_ERROR_ERRNO);
+
+#if defined(__linux__)
+    {
+        struct sockaddr_nl sa;
+
+        memset(&sa, 0, sizeof(sa));
+        sa.nl_family = AF_NETLINK;
+        sa.nl_groups = RTMGRP_LINK | RTMGRP_IPV6_IFADDR;
+        VerifyOrDie(bind(sNetlinkFd, reinterpret_cast<struct sockaddr *>(&sa), sizeof(sa)) == 0, OT_EXIT_ERROR_ERRNO);
+    }
+#endif
+
+#if defined(__APPLE__) || defined(__NetBSD__) || defined(__FreeBSD__)
+    {
+        int status;
+#ifdef ROUTE_FILTER
+        unsigned int msgfilter = ROUTE_FILTER(RTM_IFINFO) | ROUTE_FILTER(RTM_NEWADDR) | ROUTE_FILTER(RTM_DELADDR) |
+                                 ROUTE_FILTER(RTM_NEWMADDR) | ROUTE_FILTER(RTM_DELMADDR);
+#define FILTER_CMD ROUTE_MSGFILTER
+#define FILTER_ARG msgfilter
+#define FILTER_ARG_SZ sizeof(msgfilter)
```

Comment:
Are `FILTER_ARG_SZ` and `FILTER_ARG` the same under both cases?

Reply:
Yes, but that might not always be the case, if the platform's routing filter mechanism was different.